deliaioanac/assignment2/printAncestors.cpp

```diff
@@ -0,0 +1,100 @@
+#include <iostream>
+#include <vector>
+
+// Structure for a tree node
+struct TreeNode {
+    int val;
+    TreeNode *left;
+    TreeNode *right;
+    TreeNode *parent;
+    TreeNode(int x) : val(x), left(nullptr), right(nullptr), parent(nullptr) {}
+    TreeNode(int x, TreeNode *p) : val(x), left(nullptr), right(nullptr), parent(p) {}
+    ~TreeNode() {
+        delete left;
+        delete right;
+    }
+};
+
+// Class for binary tree
+class BinaryTree {
```

Comment:
I find it difficult to understand your design choices. Why do you have `createTree` instead of a constructor and `destroyTree` instead of a destructor?

I would suggest re-implementing the class as follows:

```
class BinaryTree {
    private:
        ....
    public:
        BinaryTree(const std::vector<int> &values, int k = 0, TreeNode *parent = nullptr);
        ~BinaryTree();
       void printAncestors(const int val);
};
```

Class objects are typically initialised via constructors, that't their sole purpose and you should follow this pattern unless you have a strong reason not to do so. It's also a lot better to do memory clean-up in destructors because a destructor is called whenever your object goes out of scope, meaning you don't have to remember to call it every time you're done with the object.

Also, the parameter list `...(const std::vector<int> &values, int k = 0, TreeNode *parent = nullptr);` looks messy. You never actually set k and parent to anything other than the default values when using this function in tests, so maybe it's better to never expose them as parameters at all? The caller shouldn't have to think about these values. Your code could look something like:

```
// public:
BinaryTree(const std::vector<int> &values) {
    this->tree = createTreeHelper(values, 0, nullptr);
}

// private:
TreeNode* createTreeHelper(const std::vector<int> &values, int k, TreeNode *parent) {
    ** whatever is currently in createTree **
}
```

Reply:
Also, consider having a different naming style for class members than for regular function variables.

At Google, we use a [trailing underscore](https://google.github.io/styleguide/cppguide.html#Variable_Names), so `tree` would become `tree_`. This is nice because just with a quick glance you can see where and how you're modifying class state.

Reply:
Done, done and done! I'm sorry for the poor design choices, I should have thought it through more, I was too focused on the solution and I lost sight of the design.

Reply:
Don't worry, that's what code reviews are for :) When programming you always tend to focus on the problem you need to solve, so it's good to have someone looking over your code.

It works both ways, actually. Often code reviews are also there so that the reviewer can learn something. Either coding style or to better understand the project. I was never really strong at algorithms at university, so I'm learning a lot from your cool algorithm implementations :)